vars/copyContainer.groovy

```diff
@@ -10,34 +10,67 @@
  */
 void call(Map args = [:]) {
 
+    recursive_copy = args.recursiveCopy ?: false
+    source_image = args.sourceImage
+    source_image_no_tag = source_image.split(':')[0]
+    source_registry = args.sourceRegistry
+    destination_image = args.destinationImage
+    destination_image_no_tag = destination_image.split(':')[0]
+    destination_registry = args.destinationRegistry
 
     if (args.destinationRegistry == 'opensearchstaging' || args.destinationRegistry == 'opensearchproject') {
         def dockerJenkinsCredential = args.destinationRegistry == 'opensearchproject' ? "jenkins-production-dockerhub-credential" : "jenkins-staging-dockerhub-credential"
         withCredentials([usernamePassword(credentialsId: dockerJenkinsCredential, usernameVariable: 'DOCKER_USERNAME', passwordVariable: 'DOCKER_PASSWORD')]) {
             def dockerLogin = sh(returnStdout: true, script: "set +x && echo $DOCKER_PASSWORD | docker login --username $DOCKER_USERNAME --password-stdin").trim()
-            sh """gcrane cp ${args.sourceRegistry}/${args.sourceImage} ${args.destinationRegistry}/${args.destinationImage}; docker logout"""
+            gcraneCopy()
         }
     }
+
     if (args.destinationRegistry == 'public.ecr.aws/opensearchproject') {
         withCredentials([
             string(credentialsId: 'jenkins-artifact-promotion-role', variable: 'ARTIFACT_PROMOTION_ROLE_NAME'),
             string(credentialsId: 'jenkins-aws-production-account', variable: 'AWS_ACCOUNT_ARTIFACT')])
             {
                 withAWS(role: "${ARTIFACT_PROMOTION_ROLE_NAME}", roleAccount: "${AWS_ACCOUNT_ARTIFACT}", duration: 900, roleSessionName: 'jenkins-session') {
                     def ecrLogin = sh(returnStdout: true, script: "set +x && aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin ${args.destinationRegistry}").trim()
-                    sh """gcrane cp ${args.sourceRegistry}/${args.sourceImage} ${args.destinationRegistry}/${args.destinationImage}; docker logout ${args.destinationRegistry}"""
+                    gcraneCopy()
                 }
             }
     }
+
     if(args.destinationRegistry == 'public.ecr.aws/opensearchstaging') {
-            def ecrLogin = sh(returnStdout: true, script: "set +x && aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin ${args.destinationRegistry}").trim()
-            sh """gcrane cp ${args.sourceRegistry}/${args.sourceImage} ${args.destinationRegistry}/${args.destinationImage}; docker logout ${args.destinationRegistry}"""
+        def ecrLogin = sh(returnStdout: true, script: "set +x && aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin ${args.destinationRegistry}").trim()
+        gcraneCopy()
     }
+
+}
+
+def gcraneCopy() {
+    sh """
+        set +x
+
+        if [ ${recursive_copy} = true ]; then
```

Comment:
Have limited understanding of how groovy deals with variable scope so my only concern is that is the variables declared in call() scope accessible in gcraneCopy() scope? @peterzhuamazon. 
If you have already tested this then I am fine.

Reply:
It is global variable so `gcraneCopy()` can easily access the variable defined in `call()`.
We could probably make a smaller scope access with parameter passing, but that can be a later improvement.
Thanks.

Reply:
And yes it is tested in staging.